tokio/src/net/windows/named_pipe.rs

```diff
@@ -0,0 +1,1305 @@
+/// //
+/// // Here we also make use of `first_pipe_instance`, which will ensure
+/// // that there are no other servers up and running already.
+/// let mut server = NamedPipeOptions::new()
+///     .first_pipe_instance(true)
+///     .create(PIPE_NAME)?;
+///
+/// let client = tokio::spawn(async move {
+///     // Wait forever until a socket is available to be connected to.
+///     let mut client = loop {
+///         match NamedPipeClientOptions::new().create(PIPE_NAME) {
+///             Ok(client) => break client,
+///             Err(e) if e.raw_os_error() == Some(winerror::ERROR_PIPE_BUSY as i32) => (),
+///             Err(e) => return Err(e),
+///         }
+///
+///         tokio::task::spawn_blocking(move || {
+///             wait_named_pipe(PIPE_NAME, Some(Duration::from_millis(500)))
+///         }).await??;
+///     };
+///
+///     let mut buf = [0u8; 4];
+///     client.read_exact(&mut buf[..]).await?;
+///     Ok::<_, io::Error>(buf)
+/// });
+///
+/// let server = tokio::spawn(async move {
+///     tokio::time::sleep(Duration::from_millis(200)).await;
+///
+///     // Calling `connect` is necessary for the waiting client to wake up,
+///     // even if the server is created after the client.
+///     server.connect().await?;
+///
+///     server.write_all(b"ping").await?;
+///     Ok::<_, io::Error>(())
+/// });
+///
+/// let (client, server) = tokio::try_join!(client, server)?;
+/// let payload = client?;
+/// assert_eq!(&payload[..], b"ping");
+/// let _ = server?;
+/// # Ok(()) }
+/// ```
+///
+/// # Panics
+///
+/// Panics if the specified duration is larger than `0xffffffff`
+/// milliseconds, which is roughly equal to 1193 hours.
+///
+/// ```should_panic
+/// use std::time::Duration;
+/// use tokio::net::windows::wait_named_pipe;
+///
+/// const PIPE_NAME: &str = r"\\.\pipe\mynamedpipe";
+///
+/// # fn main() -> std::io::Result<()> {
+/// wait_named_pipe(PIPE_NAME, Some(Duration::from_millis(0xffffffff)))?;
+/// # Ok(()) }
+/// ```
+pub fn wait_named_pipe(addr: impl AsRef<OsStr>, timeout: Option<Duration>) -> io::Result<()> {
+    let addr = encode_addr(addr);
+
+    let timeout = match timeout {
+        Some(timeout) => {
+            let timeout = timeout.as_millis();
+            assert! {
+                timeout < NMPWAIT_WAIT_FOREVER as u128,
+                "timeout out of bounds, can wait at most {}ms, but got {}ms", NMPWAIT_WAIT_FOREVER - 1,
+                timeout
+            };
+            timeout as DWORD
+        }
+        None => NMPWAIT_WAIT_FOREVER,
+    };
+
+    // Safety: There's nothing unsafe about this.
+    let result = unsafe { namedpipeapi::WaitNamedPipeW(addr.as_ptr(), timeout) };
+
+    if result == FALSE {
+        return Err(io::Error::last_os_error());
+    }
+
+    Ok(())
+}
+
+/// Encode an address so that it is a null-terminated wide string.
+fn encode_addr(addr: impl AsRef<OsStr>) -> Box<[u16]> {
+    addr.as_ref().encode_wide().chain(Some(0)).collect()
```

Comment:
The current code will allocate several times.
```suggestion
    let len = addr.as_ref().encode_wide().count();
    let vec = Vec::with_capacity(1 + len);
    vec.extend(addr.as_ref().encode_wide());
    vec.push(0);
    vec.into_boxed_slice()
```

Reply:
As an alternative to counting, we can use the upper bound from the `EncodeWide` iterator and fill any excess capacity with zeroes. It's still valid if the buffer has extra data after the zero-terminator.

Reply:
I'm think I'm missing something. Where does the extra allocations occur outside of `collect`?

Reply:
Oh. I think I see what you mean.

So I believe collect will make use of `size_hint` when determining the size of the buffer being used, and both of the iterators being chained supports usable size hints so the buffer *should* be appropriately sized.

Reply:
The `size_hint` of `EncodeWide` is not exact.

Reply:
Correct. It actually seems to div by 4 for some reason (instead of 2 which I'd expect) for the lower bound, so you'd usually face at least two allocations with the existing hint 🤔

Not sure how much we should care about it, so decide away. If we really want to avoid most allocations, the `PipeName` pattern as discussed earlier would be needed.

Reply:
> so decide away

I say we use the version I suggested at the top of this conversation then. We can always change it later.